### ISSUE TITLE
Attempt 4 with VI

### ISSUE THREAD

**Issue by luiarthur** [luiarthur/cytof5#13]
- [ ] Don't do the ADVI approach. Using priors and Q in same family, will yield closed form KL(q(z)||p(z)). Avoid the jacobians.
- [ ] [Gumbel softmax](https://github.com/dev4488/VAE_gumble_softmax/blob/master/vae_gumbel_softmax.py)
    - [ ] one hot representation for categorical stuff
    - [ ] implement forward & backward methods. But how?
        - [ ] Check out [autograd functions](https://pytorch.org/docs/stable/autograd.html)
- [ ] There may be better ways of dealing with Dirichlet parameters. SBT might not be the best. Again, don't do ADVI, avoid the jacobians.
    - [ ] for example, if w ~ Dir_K(a), then the variational parameters should be K-1 dimensional.
- [ ] The other way of writing KL:
    - [ ] ELBO(phi) = E[log p(x|z)] − KL(q(z) || p(z)). Keep a library of p(z). Choose q(z) to be same family. Then, KL(q(z) || p(z)) will be available in closed form. And MC samples only required for first term.
    - [ ] Note that `torch.distributions.kl.kl_divergence(p, q)` is available. Docs [here.](https://pytorch.org/docs/stable/distributions.html#module-torch.distributions.kl)
    - [ ] Note that many KL divergence pairs are [implemented](https://pytorch.org/docs/stable/_modules/torch/distributions/kl.html#kl_divergence).
- [ ] If nan appears in gradient. Try to understand what is happening, and fix them. Maybe rough empirical initial values are needed.  A quick and dirty fix is to change gradient to 0, but shouldn't be the permanent fix. Can also try using smaller learning rate.

**Comment by luiarthur** [comment]
Done. See [here](https://github.com/luiarthur/CytofPy).
